### PR TITLE
Data mover micro service node agent resume

### PR DESCRIPTION
**Pull request by Lyndon-Li**
According to design https://github.com/vmware-tanzu/velero/pull/7576, after node-agent restarts, if a DU/DD is in InProgress status, re-capture the data mover ms pod and continue the execution

### CODE REVIEW

File: pkg/builder/data_download_builder.go [hunk 1]
```diff
@@ -122,3 +123,21 @@ func (d *DataDownloadBuilder) StartTimestamp(startTime *metav1.Time) *DataDownlo
...
 	return d
 }
+
+// CompletionTimestamp sets the DataDownload's StartTimestamp.
+func (d *DataDownloadBuilder) CompletionTimestamp(completionTimestamp *metav1.Time) *DataDownloadBuilder {
+	d.object.Status.CompletionTimestamp = completionTimestamp
+	return d
+}
+
+// Labels sets the DataDownload's Progress.
```
[the remark this answers is not in the document]
Comment (Lyndon-Li) [reply pkg/builder/data_download_builder.go]: Done

File: pkg/builder/data_upload_builder.go [hunk 2]
```diff
@@ -133,7 +133,14 @@ func (d *DataUploadBuilder) Labels(labels map[string]string) *DataUploadBuilder
 	return d
 }
 
+// Labels sets the DataUpload's Progress.
```
[the remark this answers is not in the document]
Comment (Lyndon-Li) [reply pkg/builder/data_upload_builder.go]: Done

File: pkg/cmd/cli/nodeagent/server.go [hunk 3]
```diff
@@ -293,17 +293,29 @@ func (s *nodeAgentServer) run() {
...
+
+		if err := dataUploadReconciler.AttemptDataUploadResume(s.ctx, s.mgr.GetClient(), s.logger.WithField("node", s.nodeName), s.namespace); err != nil {
+			s.logger.WithError(errors.WithStack(err)).Error("failed to attempt data upload resume")
+		}
+
+		if err := dataDownloadReconciler.AttemptDataDownloadResume(s.ctx, s.mgr.GetClient(), s.logger.WithField("node", s.nodeName), s.namespace); err != nil {
+			s.logger.WithError(errors.WithStack(err)).Error("failed to attempt data download resume")
+		}
+
+		s.logger.Info("Complete to resume dataUploads and dataDownloads")
```
[the remark this answers is not in the document]
Comment (Lyndon-Li) [reply pkg/cmd/cli/nodeagent/server.go]: Done

File: pkg/controller/data_upload_controller.go [hunk 4]
```diff
@@ -890,54 +835,145 @@ func isDataUploadInFinalState(du *velerov2alpha1api.DataUpload) bool {
...
+	}()
+
+	if err := asyncBR.Init(ctx, nil); err != nil {
+		return errors.Wrapf(err, "error to init asyncBR watcher for du %s", du.Name)
+	}
+
+	if err := asyncBR.StartBackup(datapath.AccessPoint{
+		ByPath: res.ByPod.VolumeName,
+	}, du.Spec.DataMoverConfig, nil); err != nil {
+		return errors.Wrapf(err, "error to resume asyncBR watche for du %s", du.Name)
```
[the remark this answers is not in the document]
Comment (Lyndon-Li) [reply pkg/controller/data_upload_controller.go]: Done

File: pkg/controller/data_download_controller.go [hunk 5]
```diff
@@ -792,56 +737,139 @@ func isDataDownloadInFinalState(dd *velerov2alpha1api.DataDownload) bool {
...
+	}()
+
+	if err := asyncBR.Init(ctx, nil); err != nil {
+		return errors.Wrapf(err, "error to init asyncBR watcher for dd %s", dd.Name)
+	}
+
+	if err := asyncBR.StartRestore(dd.Spec.SnapshotID, datapath.AccessPoint{
+		ByPath: res.ByPod.VolumeName,
+	}, nil); err != nil {
+		return errors.Wrapf(err, "error to resume asyncBR watche for dd %s", dd.Name)
```
[the remark this answers is not in the document]
Comment (Lyndon-Li) [reply pkg/controller/data_download_controller.go]: Done

**Review by shubham-pampattiwar — APPROVED** [review]

**Review by blackpiglet — APPROVED** [review]

**Pull request merged**